### PR TITLE
[Dev] A Guide to Reproduce DeepSeek-V3 Pre-training Performance on GB200

### PR DESCRIPTION
**Pull request by yaox12**
# What does this PR do ?

Following the steps in this guide, we're able to achieve ~960 TFLOPS/GPU with 256 GB200 GPUs, which is slightly lower than the best we have internally (~970 TFLOPS). We're still investigating the cause.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: docs/discussions/deepseek-v3-gb200-optimization/deepseek-v3-gb200-reproduce-guide.md [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+--multi-latent-attention \
+--attention-dropout 0.0 \
+--hidden-dropout 0.0 \
+--clip-grad 1.0 \
+--weight-decay 0.1 \
+--qk-layernorm \
+--lr-decay-samples 584765624 \
+--lr-warmup-samples 1536000 \
+--lr-warmup-init 3.9e-7 \
+--lr 3.9e-6 \
```
Comment (Skylion007) [inline docs/discussions/deepseek-v3-gb200-optimization/deepseek-v3-gb200-reproduce-guide.md:206]: Um, isn't this a post-training LR? Odd for a PR that is suppose to be a pretraining config.

Comment (Skylion007) [reply docs/discussions/deepseek-v3-gb200-optimization/deepseek-v3-gb200-reproduce-guide.md:206]: Not sure if this config will be as stable with EP32 with a higher LR and training form scratch: may OOM due to expert imbalance.

**Comment by yanring** [comment]
+ @sbhavani for viz

File: docs/discussions/deepseek-v3-gb200-optimization/deepseek-v3-gb200-reproduce-guide.md (hunk 1 above)
Comment (yaox12) [reply docs/discussions/deepseek-v3-gb200-optimization/deepseek-v3-gb200-reproduce-guide.md:206]: We're using `--moe-router-force-load-balancing` for performance benchmarking. So these configs don't impact the performance.

Comment (yanring) [reply docs/discussions/deepseek-v3-gb200-optimization/deepseek-v3-gb200-reproduce-guide.md:206]: This guide provides a performance reference to illustrate what can be achieved under converged load balancing.

In practice, actual pretraining configurations are more complex. For instance, early training phases may involve additional recomputation or offloading to address imbalance, as well as batch size warmup during the first 469B tokens.

In short, we recommend starting from this configuration to verify that performance can be reliably reproduced and that both software and hardware are functioning correctly. After confirming stability, users can adjust the configuration for real pre-training or post-training.

**Comment by sbhavani** [comment]
Could we place this under a directory in `docs/discussions/` this will be cleaner as we add more guides and add a README

File: docs/discussions/deepseek-v3-gb200-optimization/deepseek-v3-gb200-optimization.md
Comment (yanring) [inline docs/discussions/deepseek-v3-gb200-optimization/deepseek-v3-gb200-optimization.md:1]: Thanks for the great work! Could you also ping these two in the root README?

Comment (yaox12) [reply docs/discussions/deepseek-v3-gb200-optimization/deepseek-v3-gb200-optimization.md:1]: Done.
